File: tests/typesafety/python/inline_retry.py
# expect: passes
# title: The retry loop lives inside the business function
def fetch_greeting(attempts_left: int = 3) -> str:
    failures = 0
    while True:
        try:
            return unreliable_fetch().upper()
        except ConnectionError:
            failures += 1
            if failures >= attempts_left:
                raise


def unreliable_fetch() -> str:
    raise ConnectionError("try again")

File: tests/typesafety/typescript/inline_retry.ts
// expect: passes

class ConnectionLost extends Error {}

function unreliableFetch(): string {
  throw new ConnectionLost("try again");
}

export function fetchGreeting(attemptsLeft = 3): string {
  let failures = 0;
  for (;;) {
    try {
      return unreliableFetch().toUpperCase();
    } catch (error) {
      if (!(error instanceof ConnectionLost)) throw error;
      failures += 1;
      if (failures >= attemptsLeft) throw error;
    }
  }
}
